subsys/bluetooth/host/classic/a2dp.c

```diff
@@ -99,27 +99,41 @@ static struct bt_a2dp *get_new_connection(struct bt_conn *conn)
 		return NULL;
 	}
 
+	A2DP_LOCK();
 	/* Find a space */
 	for (i = 0; i < CONFIG_BT_MAX_CONN; i++) {
```

Comment:
You're making this much more complicated than it needs to be. Since you have an array of contexts with `CONFIG_BT_MAX_CONN` elements, you can simply create a 1:1 mapping from `bt_conn` to that array with the help of `bt_conn_index()`: https://docs.zephyrproject.org/latest/doxygen/html/group__bt__conn.html#gad4aed76b80cc815f748ad0e84ae3d87c

I.e. something like:
```c
static struct bt_a2dp *get_new_connection(struct bt_conn_conn)
{
	return &connection[bt_conn_index(conn)];
}
```

I don't think you'll need the `A2DP_LOCK()` stuff after that either.

Reply:
Theoretically more than one A2DP L2CAP connections can be created based on one ACL connection. Then they are not 1:1 mapping. Usually there is only one A2DP L2CAP connection for one ACL connection. If we need to support it, then the 1:1 mapping cannot be used.

Reply:
Yes, I know, but can't you just make the general decision (at least for now) that you only support a single A2DP connection per ACL? It seems like you're already doing it, since the code is simply returning the same context if it already exists:
```
	if (connection[i].used == 1U && connection[i].session.br_chan.chan.conn == conn) {
		LOG_DBG("Conn already exists");
		A2DP_UNLOCK();
		return &connection[i];
	}
```

Reply:
I understand you, but I don't think it is good to limit the A2DP function in order to let the codes simple. I can fix the above codes that you mention. Is it OK?

Reply:
I'm struggling to understand your point

1. What I'm proposing is not adding any new limits over what you already have in the PR
2. I don't see how you could have multiple "sessions" (i.e. multiple signaling channels) over a single ACL, since the AVDTP spec doesn't define anything like that (AFAIK). It defines the initial signaling channel and how meaning is applied to subsequent L2CAP channels using the AVDTP PSM. There can be at most signaling + 3 other channels according to AVDTP spec 1.3 section 5.4.6. Also see section 1.2 where it only talks about the signaling channel in singular rather than plural.
3. Even if it was theoretically possible to have multiple signaling channels for an ACL. I'm not aware of any devices that do that. Are you? Adding code complexity for non-existent theoretical situation is not a good practice IMO.
4. The array of contexts you have uses `CONFIG_BT_MAX_CONN`, implying that you really do want a 1:1 relationship. Otherwise it'd make more sense to define a new Kconfig option for this.

Reply:
OK, @jhedberg Thanks, I see the follow line in the spec too. I will update the codes.
> The signaling channel (1) is only established once between two streaming devices.

Reply:
updated. please help to review again.